### ISSUE TITLE
Tag Suggestions

### ISSUE THREAD

**Issue by sco1** [python-discord/meta#22]
PyDis' Python bot uses `!tags` to provide brief, canonical explanations for common questions, programming pitfalls, and other Python related topics.

Use this issue to suggest additional tags, bonus points if you help out with the body of the tag as well!

**Comment by sco1** [comment]
# Heroku

With its free tier, Heroku is a great resource for getting started with a variety of different web-based Python projects. However, the free tier comes with some limitations that make it not scale particularly well, or make it not well suited for certain applications (like discord bots).

This tag should expand on these limitations, and also link to the VPS section of our Resources page ([pending](https://github.com/python-discord/site/issues/177)) for more info and some possible alternatives for growing your hosting.

**Comment by heavysaturn** [comment]
These were suggested back on GitLab, and were never written. If anyone would like write something up for these, we'd be happy for the contribution.

- [ ]  async
- [ ]  decorators
- [ ]  dict comprehensions
- [ ]  staticmethod

**Comment by gdude2002** [comment]
My suggestion is an explainer on why `discord.py` isn't suitable for beginners, and what people need to know in order to work effectively with it.

I'm not sure what to use as the tag name. `discord-is-hard`? Something like that? And then maybe the content could be along the lines of..

```md
Before talking a Discord bot using the `discord.py` library, please bear in mind that **`discord.py` is not beginner-friendly**. We recommend that anyone attempting to work with it first attempts to gain some intermediate Python experience.

We recommend that all users attempting to work with `discord.py` are familiar with the following: **asyncio**, **classes**, **decorators**, **pip**, and **how to read Python documentation**.

If you're a beginner and you'd like some help getting started with Python, then feel free to check out our [resources page](https://pythondiscord.com/info/resources) - where we've compiled a list of goodies that we regularly recommend to aspiring Python developers.
```

I admit I'm not 100% confident that that list encompasses everything. Thoughts?

**Comment by scragly** [comment]
Rather than a codeblock of markdown, I feel it's easier to get a feel for it's appearance and how it reads if it was just normal styled text. I don't get distracted by having to replace styles in my head this way, so I've also put it here for reference.

> Before talking a Discord bot using the `discord.py` library, please bear in mind that **`discord.py` is not beginner-friendly**. We recommend that anyone attempting to work with it first attempts to gain some intermediate Python experience.
> 
> We recommend that all users attempting to work with `discord.py` are familiar with the following: **asyncio**, **classes**, **decorators**, **pip**, and **how to read Python documentation**.
> 
> If you're a beginner and you'd like some help getting started with Python, then feel free to check out our [resources page](https://pythondiscord.com/info/resources) - where we've compiled a list of goodies that we regularly recommend to aspiring Python developers.

"Before talking a Discord bot" sounds weird, and also sweeps across other non-discord.py projects which we don't really touch on for the moment. "Before working on a bot with `discord.py`" may be preferred here.

The list doesn't encompass everything, but that's not the purpose. What you have touches on the more commonly lacking aspects that people don't think to address; the most common issues I've seen being hit are lacking knowledge of classes/instances, asyncio concepts, namespace/scope and interpreter environments (such as installing with `pip install discord.py` and having it show as not installed when running the script with `py scriptname.py`). Maybe we could direct each point to their respective official documentation/tutorial also.

For more extensive listings of required knowledge, I do have a section in my guide here:
https://gist.github.com/scragly/095b5278a354d46e86f02d643fc3d64b#required-knowledge
Eventually, I'll be migrating and expanding this (currently basic) guide when we have the django wiki up.

One thing I want to say though is that there's quite a few people whose first exposure to Python is through working with discord.py projects. I myself am included in that. I really hope that such a tag as this doesn't end up being overused, or the default response whenever someone doesn't know something. In fact, if there was even a slim chance of this happening, I'd prefer that this tag simply did not exist, and leaving it to normal human responses to address cases like this.

**Comment by gdude2002** [comment]
:laughing: 

> Before talking a Discord bot

Totally my fault, I meant to say "Before working on"!

As discussed in the meta channel, it seems we all agree that a more friendly and lengthy FAQ is a better option. Given that, should we wait for the Django rewrite, or update the FAQ before that happens?

**Comment by heavysaturn** [comment]
the FAQ hasn't been ported to Django yet (that would be my job!), so I think it's okay to PR this in before that happens.

**Comment by fiskenslakt** [comment]
Today I was asked to explain the purpose of `if __name__ == '__main__':` and after typing out the explanation, I wondered if it might be worth having a tag for this.

The explanation is simple, but to be explained thoroughly, it requires a fair amount of typing. For example, after explaining how `__name__` works, demonstrate by showing 2 files, one where `__name__` is printed, and another where that file is imported.

**Comment by GhostofGoes** [comment]
The Python REPL is something I've explained a lot, and it would be useful to have a tag for.

Basically:
* What REPL stands for
* What the REPL console is
* How it's different from running the code in a file

**Comment by sco1** [comment]
Since there's often an overlap between what's desired for a tag vs. the ability for a wiki article to be more comprehensive, I've expanded the scope of this to also include suggestions for Wiki articles (implemented soon) so you don't have to decide whether or not this belongs as a tag or as a Wiki article.

**Comment by fiskenslakt** [comment]
### putting these here so i remember to do them
anyone is welcome to beat me to it (please :sweat:)
- [ ] generators
- [ ] mutable default arguments
- [ ] iterators
- [ ] self
- [ ] magic methods
- [ ] inheritance
- [ ] scope
- [ ] global interpreter lock
- [ ] import (where it looks, packages vs modules, `from` syntax, etc)
- [ ] mutability
- [ ] variables (how references work, link to ned's stuff about this)
- [ ] super
- [ ] bytecode (compilation, interpretation, .pyc files, `dis` module, etc)
- [ ] mutual imports (on the fence for this one, not a common topic afaik)
- [ ] freezing (pyinstaller and the like, how it works maybe, etc)
- [ ] operator precedence
- [ ] ternary
- [ ] performance (how to speed up your code, `timeit` module, `profile` module, more efficient algs, better data structures, multi-threading/processing, C extensions, Cython, pypy)
this one will obviously have to be fairly brief, mostly providing links for more information
- [ ] floating-point arithmetic (precision errors, `.1 + .2`)
- [ ] bankers rounding
- [ ] eval/exec

I <3 tags

**Comment by MarkKoz** [comment]
A common question with Qt is how to have two widgets interact with each other when they're defined in separate modules. I think the topic may be too complex for writing the explanation ourselves. I want to try to find a resource that can cover this clearly but that may be difficult to find. Maybe this is a better fit for the wiki in the future 🤔 .

**Comment by SebastiaanZ** [comment]
From the meta channel:

> **v_gz** *Today at 4:19 PM*
> May I post a suggestion for the shortcut explanations of the bot?
> I have noticed a common pitfall iamong newcomers: func() (function call) versus func (the function object itself). For instance, when it is needed to pass the function as an argument. Maybe it would be nice to have a command like !func()-vs-func that will contain a clear and concise explanation?

**Comment by gwerbin** [comment]
In the `dict-iterate` tag, should also mention
```python
for v in dict.values():
    print(v)
```

We could maybe have a more general "comprehensions" tag with general syntax? Plus a wiki entry talking in more detail about them.

**Comment by gwerbin** [comment]
Potentially useful demo for a potential `generator` or `iterator` tag (I would consider these closely-related topics and one should probably link to the other): https://repl.it/@maximum__/lazy-generator

**Comment by ItsCinnabar** [comment]
I would like a !pathlib command. Most new programmers don't know about pathlib and the fact that they should be using it over os.path (if python 3.6+). They google about paths, see legacy code using os.path, and assume that is the proper/best way to handle them. A !pathlib command could be used when we see os.path and could give a little plug about pathlib and a bit of information.


> Python 3 comes with a new module named Pathlib. Since Python 3.6, pathlib.Path objects work nearly everywhere that os.path can be used, meaning you can integrate your new code directly into legacy code without having to rewrite anything. Pathlib makes working with paths way simpler than os.path. Programming for cross platform paths is the big concern when dealing with paths, and pathlib makes it a breeze.
> 
> If you would like more information about pathlib, see the following articles : 
> https://treyhunner.com/2018/12/why-you-should-be-using-pathlib/
> https://treyhunner.com/2019/01/no-really-pathlib-is-great/
> 
> If you are ready to dive in, check out the pathlib documentation :
> https://docs.python.org/3/library/pathlib.html

**Comment by sco1** [comment]
Migrating this to the bot repository. These may be PRed once https://github.com/python-discord/bot/pull/803 is merged.
